OpenID Connect - Validate Discovery Endpoint response

cieciurm opened the pull request:
<!--  Thanks for sending a pull request!
-->

**What this PR does / why we need it**:
- This PR adds validation of OpenID Connect Discover response
- Validation follows rules from [the spec](https://openid.net/specs/openid-connect-discovery-1_0.html#IssuerDiscovery)

**Which issue(s) this PR fixes**: 
closes #701

**Special notes for your reviewer**:

- This is a very early PR
- I want to ask for feedback on the approach
- Tests will be added when we agree on solution

**Does this PR introduce a user-facing change?**:
- No

Please make sure you've completed the relevant tasks for this PR, out of the following list:

- [x] Code compiles correctly
- [x] Created/updated tests
- [x] Unit tests passing
- [x] End-to-end tests passing
- [ ] Extended the documentation
- [ ] Provided sample for the feature

cieciurm commented:
Thanks for any suggestions and feedback 😄 
cc @sungam3r

sungam3r commented on src/HealthChecks.OpenIdConnectServer/OidcValidationHelper.cs:
```diff
@@ -0,0 +1,29 @@
+        $"Invalid discover response - '{value}' must be set!";
+
+    private static string GetMissingRequiredValuesExceptionMessage(string value, IEnumerable<string> requiredValues) =>
+        $"Invalid discover response - '{value}' must contain {string.Join(",", requiredValues)}!";
```
```suggestion
        $"Invalid discover response - '{value}' must be one of the following values: {string.Join(",", requiredValues)}!";
```

sungam3r commented on src/HealthChecks.OpenIdConnectServer/OidcConstants.cs:
```diff
@@ -0,0 +1,22 @@
+
+    internal const string ALGORITHMS_SUPPORTED = "id_token_signing_alg_values_supported";
+
+    internal static IEnumerable<string> REQUIRED_RESPONSE_TYPES => new[] { "code", "id_token", "token id_token" };
+
+    internal static IEnumerable<string> REQUIRED_SUBJECT_TYPES => new[] { "pairwise", "public" };
+
+    internal static IEnumerable<string> REQUIRED_ALGORITHMS => new[] { "RS256" };
```
```suggestion
    internal static string[] REQUIRED_RESPONSE_TYPES => new[] { "code", "id_token", "token id_token" };

    internal static string[] REQUIRED_SUBJECT_TYPES => new[] { "pairwise", "public" };

    internal static string[] REQUIRED_ALGORITHMS => new[] { "RS256" };
```

cieciurm replied:
Same question as above. Here I am using only `Contains(...)` method  so I thought I would use the most most basic collection interface.

sungam3r replied:
>  so I thought I would use the most most basic collection interface.

Indeed, but it's not needed when working in internal (i.e. 100% your) code.

sungam3r commented on src/HealthChecks.OpenIdConnectServer/IdSvrHealthCheck.cs:
```diff
@@ -27,13 +28,35 @@ public async Task<HealthCheckResult> CheckHealthAsync(HealthCheckContext context
         }
     }
+
+    private static void ValidateResponse(DiscoveryEndpointResponse response)
```
Let's make it an instance method of `DiscoveryEndpointResponse` + place https://openid.net/specs/openid-connect-discovery-1_0.html#IssuerDiscovery link in comments for that method.

cieciurm replied:
Thanks, done.

sungam3r commented on src/HealthChecks.OpenIdConnectServer/DiscoveryEndpointResponse.cs:
```diff
@@ -0,0 +1,24 @@
+    public string JwksUri { get; set; } = null!;
+
+    [JsonPropertyName(OidcConstants.RESPONSE_TYPES_SUPPORTED)]
+    public IEnumerable<string> ResponseTypesSupported { get; set; } = null!;
+
+    [JsonPropertyName(OidcConstants.SUBJECT_TYPES_SUPPORTED)]
+    public IEnumerable<string> SubjectTypesSupported { get; set; } = null!;
+
+    [JsonPropertyName(OidcConstants.ALGORITHMS_SUPPORTED)]
+    public IEnumerable<string> IdTokenSigningAlgValuesSupported { get; set; } = null!;
```
```suggestion
    public string[] ResponseTypesSupported { get; set; } = null!;

    [JsonPropertyName(OidcConstants.SUBJECT_TYPES_SUPPORTED)]
    public string[] SubjectTypesSupported { get; set; } = null!;

    [JsonPropertyName(OidcConstants.ALGORITHMS_SUPPORTED)]
    public string[] IdTokenSigningAlgValuesSupported { get; set; } = null!;
```

cieciurm replied:
Could you share more thoughts why array is better than IEnumerable here?

sungam3r replied:
1. Nothing to abstract here. Internal code.
2. Abstraction gives nothing except additional iterator allocation when using `Contains` etc.

cieciurm replied:
Understood, thanks!

sungam3r commented:
> I want to ask for feedback on the approach
Tests will be added when we agree on solution

Looks good thought I have not read the spec.

sungam3r commented on src/HealthChecks.OpenIdConnectServer/DiscoveryEndpointResponse.cs:
```diff
@@ -0,0 +1,63 @@
+        }
+    }
+
+    private static void ValidateRequiredValues(IEnumerable<string> values, string metadata, IEnumerable<string> requiredValues)
```
```suggestion
    private static void ValidateRequiredValues(IEnumerable<string> values, string metadata, string[] requiredValues)
```

sungam3r commented on src/HealthChecks.OpenIdConnectServer/DiscoveryEndpointResponse.cs:
```diff
@@ -0,0 +1,63 @@
+        }
+    }
+
+    private static bool AnyValueContains(IEnumerable<string> values, IEnumerable<string> requiredValues) =>
```
```suggestion
    private static bool AnyValueContains(IEnumerable<string> values, string[] requiredValues) =>
```

sungam3r commented on src/HealthChecks.OpenIdConnectServer/DiscoveryEndpointResponse.cs:
```diff
@@ -0,0 +1,63 @@
+    private static string GetMissingValueExceptionMessage(string value) =>
+        $"Invalid discovery response - '{value}' must be set!";
+
+    private static string GetMissingRequiredValuesExceptionMessage(string value, IEnumerable<string> requiredValues) =>
```
```suggestion
    private static string GetMissingRequiredValuesExceptionMessage(string value, string[] requiredValues) =>
```

sungam3r commented on src/HealthChecks.OpenIdConnectServer/DiscoveryEndpointResponse.cs:
```diff
@@ -0,0 +1,63 @@
+        }
+    }
+
+    private static bool AnyValueContains(IEnumerable<string> values, string[] requiredValues) =>
```
```suggestion
    private static bool AnyValueContains(string[] values, string[] requiredValues) =>
```

sungam3r commented on src/HealthChecks.OpenIdConnectServer/DiscoveryEndpointResponse.cs:
```diff
@@ -0,0 +1,63 @@
+        }
+    }
+
+    private static void ValidateRequiredValues(IEnumerable<string> values, string metadata, string[] requiredValues)
```
```suggestion
    private static void ValidateRequiredValues(string[] values, string metadata, string[] requiredValues)

```

cieciurm commented:
I would like to add tests for validation.

sungam3r commented:
Reminder - bump version.

cieciurm commented:
I've added applied code suggestions, added unit tests and bumped the version.

I decided to write more sophisticated check for `response_types_supported`. The spec mentiones a combined response type `token id_token`, but some identity providers (f.e. Identity Server and Azure AD) return `id_token token`. So I decided to validate separately the single response types (`code`, `id_token`) and separately combined response type. This should prevent false negatives.

cc @sungam3r

sungam3r commented on src/HealthChecks.OpenIdConnectServer/DiscoveryEndpointResponse.cs:
```diff
@@ -0,0 +1,75 @@
+
+        ValidateRequiredValues(ResponseTypesSupported, OidcConstants.RESPONSE_TYPES_SUPPORTED, OidcConstants.REQUIRED_RESPONSE_TYPES);
+
+        // Specification decribes 'token id_token' response type,
```
```suggestion
        // Specification describes 'token id_token' response type,
```

sungam3r commented on src/HealthChecks.OpenIdConnectServer/HealthChecks.OpenIdConnectServer.csproj:
```diff
@@ -11,6 +11,11 @@
+  </ItemGroup>
+
+  <ItemGroup>
+    <InternalsVisibleTo Include="HealthChecks.OpenIdConnectServer.Tests" />
```
See #1987

cieciurm commented:
Thanks, `<InternalsVisibleTo />` has been removed. 

cc @sungam3r
